src/lib.rs

```diff
@@ -210,88 +119,324 @@ fn get_repo(pkg_info: &Value) -> Result<Option<String>, Error> {
+        work_dir: &Path,
+    ) -> Cloner {
+        Cloner {
+            registry_url: registry_url.to_string(),
+            github_url: github_url.to_string(),
+            gitlab_url: gitlab_url.to_string(),
+            bitbutcket_url: bitbutcket_url.to_string(),
+            work_dir: work_dir.into(),
+        }
     }
-    versions.sort_unstable_by_key(|x| x.1.clone());
-    let last = versions.last().unwrap().0;
-    let dl_path = last["dl_path"]
-        .as_str()
-        .expect("Could not find `dl_path` in crate version info.");
-    let dl_path = format!("https://crates.io{}", dl_path);
-    let version = last["num"]
-        .as_str()
-        .expect("Could not find `num` in crate version info.");
-    println!("Downloading `{}`", dl_path);
-    let mut response =
-        reqwest::get(&dl_path).context(format!("Failed to download `{}`", dl_path))?;
-    // TODO: This could be much better.
-    let mut body = Vec::new();
-    response.copy_to(&mut body)?;
-    let gz = GzDecoder::new(body.as_slice());
-    let mut tar = Archive::new(gz);
-    let base = format!("{}-{}", name.to_lowercase(), version);
 
-    for entry in tar.entries()? {
-        let mut entry = entry.context("Failed to get tar entry.")?;
-        let entry_path = entry
-            .path()
-            .context("Failed to read entry path.")?
-            .into_owned();
-        println!("{}", entry_path.display());
+    /// Create a Crate Cloner using all the default settings
+    pub fn default() -> std::io::Result<Cloner> {
+        Ok(Cloner {
+            registry_url: DEFAULT_REGISTRY_URL.to_string(),
+            github_url: DEFAULT_GITHUB_URL.to_string(),
+            gitlab_url: DEFAULT_GITLAB_URL.to_string(),
+            bitbutcket_url: DEFAULT_BITBUCKET_URL.to_string(),
+            work_dir: env::current_dir()?,
+        })
+    }
 
-        // Sanity check.
-        if !entry_path.starts_with(&base) {
+    /// Clones a crate using the provided method.
+    ///
+    /// - `method_kind` - Method to fetch crate.
+    /// - `spec` - The name of the crate to clone
+    /// - `version` - The semantic version (semver) of the spec crate to clone
+    /// - `extra` - Additional arguments passed to clone command.
+    ///
+    pub fn clone(
+        &self,
+        method_kind: CloneMethodKind,
+        spec: &str,
+        version: Option<&str>,
+        extra: &[&str],
+    ) -> Result<(), Error> {
+        let mut parts = spec.splitn(2, ':');
+        let name = parts.next().unwrap();
+        let spec_version_req = parts.next();
+        if spec_version_req.is_some() && version.is_some() {
+            bail!("Cannot specify both a :version and --version.");
+        }
+        let version_req = version
+            .or(spec_version_req)
+            .map(check_semver_req)
+            .transpose()?;
+        let pkg_info = self.get_pkg_info(name)?;
+        let repo = get_repo(&pkg_info)?;
+        let (method, repo) = match method_kind.command() {
```

Comment:
Perhaps this should just match on the enum?

Reply:
Done